File: src/main/javabak/datagen/DataGenerators.java
package redcrafter07.processed.datagen;

import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.data.event.GatherDataEvent;
import redcrafter07.processed.ProcessedMod;

@EventBusSubscriber(modid = ProcessedMod.ID, bus = EventBusSubscriber.Bus.MOD)
public class DataGenerators {
    @SubscribeEvent
    public static void gatherData(GatherDataEvent event) {
        final var generator = event.getGenerator();
        final var packOutput = generator.getPackOutput();
        final var lookupProvider = event.getLookupProvider();
        final var existingFileHelper = event.getExistingFileHelper();

        generator.addProvider(event.includeServer(), ModLootTableProvider.create(packOutput, lookupProvider));
        generator.addProvider(event.includeServer(), new ModWorldGenProvider(packOutput, lookupProvider));

        generator.addProvider(event.includeClient(), new ModBlockStateProvider(packOutput, existingFileHelper));
        generator.addProvider(event.includeClient(), new ModItemModelProvider(packOutput, existingFileHelper));

        final var blockTagGenerator = generator.addProvider(
            event.includeServer(), new ModBlockTagGenerator(packOutput, lookupProvider, existingFileHelper)
        );
        generator.addProvider(
            event.includeServer(),
            new ModItemTagGenerator(packOutput, lookupProvider, blockTagGenerator.contentsGetter(), existingFileHelper)
        );
    }
}

File: src/main/kotlin/redcrafter07/processed/datagen/DataGenerators.kt
package redcrafter07.processed.datagen

import net.minecraft.data.DataGenerator
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.common.data.ExistingFileHelper
import net.neoforged.neoforge.data.event.GatherDataEvent
import redcrafter07.processed.ProcessedMod

@EventBusSubscriber(modid = ProcessedMod.ID, bus = EventBusSubscriber.Bus.MOD)
object DataGenerators {
    @SubscribeEvent
    fun gatherData(ev: GatherDataEvent) {
        val generator: DataGenerator = ev.generator
        val packOutput = generator.packOutput
        val lookupProvider = ev.lookupProvider
        val existingFileHelper: ExistingFileHelper = ev.existingFileHelper

        generator.addProvider(ev.includeServer(), ModLootTableProvider.create(packOutput, lookupProvider))
        generator.addProvider(ev.includeServer(), ModWorldGenProvider(packOutput, lookupProvider))

        generator.addProvider(ev.includeClient(), ModBlockStateProvider(packOutput, existingFileHelper))
        generator.addProvider(ev.includeClient(), ModItemModelProvider(packOutput, existingFileHelper))

        val blockTagGenerator = generator.addProvider(
            ev.includeServer(), ModBlockTagGenerator(packOutput, lookupProvider, existingFileHelper)
        )
        generator.addProvider(
            ev.includeServer(),
            ModItemTagGenerator(packOutput, lookupProvider, blockTagGenerator.contentsGetter(), existingFileHelper)
        )

    }
}